### PR TITLE
testing refactoring PR

### PR DESCRIPTION
**Pull request by miguelraz**
this is a test.

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
This needs DiffBase, ForwardDiff, and NLsolve to be added. You should `using` those since they will be very common in this library. Additionally, they need to be added to the REQUIRE file.

**Comment by miguelraz** [comment]
Added a `using DiffBase, ForwardDiff, NLsolve` to DiffEqDiffTools.jl.
Added those same dependencies to REQUIRE.

**Comment by miguelraz** [comment]
Dependencies added.

File: src/DiffEqDiffTools.jl [hunk 1]
```diff
@@ -1,5 +1,74 @@
...
+end
+
+Base.@pure DiffCache(u::AbstractArray) = DiffCache(eltype(u),length(u),Val{ForwardDiff.pickchunksize(length(u))})
+Base.@pure DiffCache(u::AbstractArray,nlsolve) = DiffCache(eltype(u),length(u),Val{get_chunksize(nlsolve)})
+Base.@pure DiffCache{CS}(u::AbstractArray,T::Type{Val{CS}}) = DiffCache(eltype(u),length(u),T)
+
+get_du{T<:Dual}(dc::DiffCache, ::Type{T}) = dc.dual_du
+get_du(dc::DiffCache, T) = dc.du
+
+realtype{T}(::Type{T}) = T
```
Comment (ChrisRackauckas) [inline src/DiffEqDiffTools.jl]: `realtype` should go to DiffEqBase 

https://github.com/JuliaDiffEq/DiffEqBase.jl/blob/master/src/utils.jl

and uses of it in OrdinaryDiffEq should get qualified

Comment (ChrisRackauckas) [inline src/DiffEqDiffTools.jl]: Err actually, just define that in OrdinaryDiffEq . No reason to have that in DiffEqBase.

**Comment by ChrisRackauckas** [comment]
You should add unit tests. NLSOLVEJL_SETUP is defined here:

http://docs.juliadiffeq.org/latest/features/linear_nonlinear.html#Basic-nlsolve-mehtod:-NLSOLVEJL_SETUP-1

You should define a rootfinding problem `f(x)=0` where you know the answer, and then use the two steps.

```julia
NLSOLVEJL_SETUP()
nl_rhs = nlsolve(Val{:init},rhs,u0)
```

to initialize and

```julia
nlres = nlsolve(nl_rhs,uhold)
```

to use it. `f(nlres)=0` should hold. Test this with a few different chunksizes and autodifferentiation flipped.

Note that this is exactly how the implicit ODE solvers use it. For example, implicit Euler:

https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/master/src/caches.jl#L1134
https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/master/src/integrators/implicit_integrators.jl#L28

This will complete this PR.

File: src/DiffEqDiffTools.jl [hunk 2]
```diff
@@ -1,5 +1,74 @@
...
+
+Base.@pure function DiffCache{chunk_size}(T, length, ::Type{Val{chunk_size}})
+    DiffCache(zeros(T, length), zeros(Dual{chunk_size, T}, length))
+end
+
+Base.@pure DiffCache(u::AbstractArray) = DiffCache(eltype(u),length(u),Val{ForwardDiff.pickchunksize(length(u))})
+Base.@pure DiffCache(u::AbstractArray,nlsolve) = DiffCache(eltype(u),length(u),Val{get_chunksize(nlsolve)})
+Base.@pure DiffCache{CS}(u::AbstractArray,T::Type{Val{CS}}) = DiffCache(eltype(u),length(u),T)
+
+get_du{T<:Dual}(dc::DiffCache, ::Type{T}) = dc.dual_du
```
Comment (ChrisRackauckas) [inline src/DiffEqDiffTools.jl]: You're going to need to `import ForwardDiff: Dual` in order to use `Dual` like this. The same statement should be removed from OrdinaryDiffEq and StochasticDiffEq.

**Pull request closed without merging**